File: policies/isUS/kotlin/src/test/kotlin/com/octetproof/octetpolicy/IsUSTests.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Understone, Inc.

package com.octetproof.octetpolicy

import com.octetproof.sdk.api.ConfidenceSummary
import com.octetproof.sdk.api.LocationProof
import com.octetproof.sdk.api.OctetRegion
import com.octetproof.sdk.api.OctetVerdict
import com.octetproof.sdk.api.ProofLevel
import com.octetproof.sdk.api.ProofRegion
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.time.Instant

/**
 * Tests for [isUS] against the region-containment model. The
 * territory exclusion is now the SDK's responsibility (PR/GU/etc.
 * yield NO from country("US")), so it's no longer pinned here.
 * Mirrors the Swift IsUSTests.
 */
class IsUSTests {

    private val fixedTime: Instant = Instant.ofEpochSecond(1_716_000_000)

    // 0.0.1-alpha LocationProof has no default instance; tests don't read
    // proof bytes, so synthesize a placeholder with arbitrary fields.
    private fun fakeProof(): LocationProof = LocationProof(
        "test", ProofRegion.Country("US"), ProofLevel.COUNTRY, 0L,
        ConfidenceSummary(), ByteArray(0), "test", "test", ByteArray(0),
    )

    private fun verdict(
        result: OctetVerdict.Result,
        reason: OctetVerdict.ReasonCode = OctetVerdict.ReasonCode.OK,
    ): OctetVerdict {
        val proof: LocationProof? =
            if (result == OctetVerdict.Result.INDETERMINATE) null
            else fakeProof()
        return OctetVerdict(
            result = result, reason = reason, message = "test",
            proof = proof, validity = null, queriedAt = fixedTime,
            confidence = ConfidenceSummary(),
        )
    }

    @Test
    fun `device in US matches`() = runTest {
        val result = isUS { region ->
            assertEquals(OctetRegion.country("US"), region)
            verdict(OctetVerdict.Result.YES)
        }
        assertTrue(result.match)
        assertEquals("US", result.country)
        assertNull(result.state)
        assertEquals(PolicyReasonCode.OK, result.reason)
        assertEquals(Confidence.HIGH, result.confidence)
        assertEquals("isUS", result.policyName)
        assertEquals(PolicyPackage.VERSION, result.policyVersion)
        assertEquals(fixedTime, result.evaluatedAt)
    }

    @Test
    fun `device outside US is clean negative`() = runTest {
        val result = isUS { verdict(OctetVerdict.Result.NO) }
        assertFalse(result.match)
        assertNull(result.country,
            "a NO verdict doesn't reveal where the device is — country stays null")
        assertEquals(PolicyReasonCode.OK, result.reason)
    }

    @Test
    fun `indeterminate verdict is unknown not negative`() = runTest {
        val result = isUS {
            verdict(OctetVerdict.Result.INDETERMINATE, OctetVerdict.ReasonCode.NO_FIX)
        }
        assertFalse(result.match)
        assertNull(result.country)
        assertEquals(PolicyReasonCode.VERDICT_INDETERMINATE, result.reason,
            "match=false here means 'we cannot tell', not 'device is not in the US'")
    }
}

File: policies/isUS/swift/Tests/IsUSTests.swift
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Understone, Inc.

import XCTest
import OctetSDK
@testable import IsUS
@testable import OctetPolicyCore

final class IsUSTests: XCTestCase {

    private let fixedTime = Date(timeIntervalSince1970: 1_716_000_000)

    // 0.0.1-alpha LocationProof has no default init; tests don't read
    // proof bytes, so synthesize a placeholder with arbitrary fields.
    private func fakeProof() -> LocationProof {
        LocationProof(
            id: "test", claimedRegion: .country(isoCode: "US"),
            level: .country, timestampMs: 0,
            confidence: ConfidenceSummary(),
            positionCommitment: Data(), sdkVersion: "test",
            platform: "test", proofBytes: Data()
        )
    }

    private func verdict(
        _ result: OctetVerdict.Result,
        reason: OctetVerdict.ReasonCode = .ok
    ) -> OctetVerdict {
        let proof: LocationProof? = (result == .indeterminate) ? nil : fakeProof()
        return OctetVerdict(
            result: result, reason: reason, message: "test",
            proof: proof, validity: nil, queriedAt: fixedTime,
            confidence: ConfidenceSummary()
        )
    }

    func test_device_in_us_matches() async {
        let result = await isUS { region in
            XCTAssertEqual(region.shape, .country(isoCode: "US"),
                           "isUS must query the US country region")
            return self.verdict(.yes)
        }
        XCTAssertTrue(result.match)
        XCTAssertEqual(result.country, "US")
        XCTAssertNil(result.state)
        XCTAssertEqual(result.reason, .ok)
        XCTAssertEqual(result.policyName, "isUS")
        XCTAssertEqual(result.policyVersion, PolicyPackage.version)
    }

    func test_device_outside_us_is_clean_negative() async {
        // A device in a territory (PR, GU, …) or any other country
        // yields .no from .country("US") — the SDK handles the
        // territory exclusion. Clean negative, reason .ok.
        let result = await isUS { _ in self.verdict(.no) }
        XCTAssertFalse(result.match)
        XCTAssertNil(result.country)
        XCTAssertEqual(result.reason, .ok)
    }

    func test_indeterminate_is_unknown_not_negative() async {
        let result = await isUS { _ in self.verdict(.indeterminate, reason: .noFix) }
        XCTAssertFalse(result.match)
        XCTAssertEqual(result.reason, .verdictIndeterminate)
    }
}
